modules/apps/site-navigation/site-navigation-service/src/main/java/com/liferay/site/navigation/internal/upgrade/v3_0_0/SiteNavigationMenuItemExternalReferenceCodeUpgradeProcess.java

```diff
@@ -0,0 +1,97 @@
+/**
+ * SPDX-FileCopyrightText: (c) 2025 Liferay, Inc. https://liferay.com
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR LicenseRef-Liferay-DXP-EULA-2.0.0-2023-06
+ */
+
+package com.liferay.site.navigation.internal.upgrade.v3_0_0;
+
+import com.liferay.portal.kernel.dao.jdbc.AutoBatchPreparedStatementUtil;
+import com.liferay.portal.kernel.model.ExternalReferenceCodeModel;
+import com.liferay.portal.kernel.model.PersistedModel;
+import com.liferay.portal.kernel.upgrade.UpgradeProcess;
+import com.liferay.portal.kernel.util.UnicodeProperties;
+import com.liferay.portal.kernel.util.UnicodePropertiesBuilder;
+import com.liferay.site.navigation.menu.item.util.SiteNavigationMenuItemUtil;
+import com.liferay.site.navigation.type.SiteNavigationMenuItemType;
+import com.liferay.site.navigation.type.SiteNavigationMenuItemTypeRegistry;
+
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+
+/**
+ * @author Joao Victor Alves
+ */
+public class SiteNavigationMenuItemExternalReferenceCodeUpgradeProcess
+	extends UpgradeProcess {
+
+	public SiteNavigationMenuItemExternalReferenceCodeUpgradeProcess(
+		SiteNavigationMenuItemTypeRegistry siteNavigationMenuItemTypeRegistry) {
+
+		_siteNavigationMenuItemTypeRegistry =
+			siteNavigationMenuItemTypeRegistry;
+	}
+
+	@Override
+	protected void doUpgrade() throws Exception {
+		try (PreparedStatement preparedStatement1 = connection.prepareStatement(
+				"select ctCollectionId, siteNavigationMenuItemId, type_, " +
+					"typeSettings from SiteNavigationMenuItem");
+			PreparedStatement preparedStatement2 =
+				AutoBatchPreparedStatementUtil.concurrentAutoBatch(
+					connection,
+					"update SiteNavigationMenuItem set typeSettings = ? " +
+						"where ctCollectionId = ? and " +
+							"siteNavigationMenuItemId = ?");
+
+			ResultSet resultSet = preparedStatement1.executeQuery()) {
+
+			while (resultSet.next()) {
+				String navigationMenuItemType = resultSet.getString("type_");
+
+				if (!SiteNavigationMenuItemUtil.isExternalReferenceCodeType(
+						navigationMenuItemType)) {
+
+					continue;
+				}
+
+				UnicodeProperties typeSettingsUnicodeProperties =
+					UnicodePropertiesBuilder.fastLoad(
+						resultSet.getString("typeSettings")
+					).build();
+
+				SiteNavigationMenuItemType siteNavigationMenuItemType =
+					_siteNavigationMenuItemTypeRegistry.
+						getSiteNavigationMenuItemType(navigationMenuItemType);
+
+				PersistedModel model = siteNavigationMenuItemType.getModel(
+					typeSettingsUnicodeProperties);
+
+				if (model instanceof ExternalReferenceCodeModel) {
```

Comment:
I believe we can be more aggressive with not updating unnecessary models, see https://github.com/jonathanmccann/liferay-portal/commit/03d360b27db1be525e86bfaa246b1195d758a4c3